draft-trammell-taps-post-sockets.md

```diff
@@ -807,7 +816,7 @@ Local and Remote, depending on how it was configured.
 // open a connection to a server using an existing Association and send some data,
 // which will be sent early if possible.
 func sayHelloWithAssociation(association Association) {
-    carrier := InitiateWithAssociation(association)
+    carrier := association.GetCarrier()
```

Comment:
So I assume that getCarrier is initiating a new carrier, using the existing association state. This may be the bias of languages I’m used to, but “get” feels like accessing a reference to something that already exists.

Reply:
+1 -- maybe rename to CreateCarrier

Reply:
The intent here is that it returns a carrier if one is already available, creating one if not, in keeping with the "application doesn't care about carrier internal state" design principle... however, yeah, that design principle leads to kind of silly naming, because what it *really* does is GetOrCreateCarrier

Reply:
...maybe association.GetActiveCarrier() ...? Or call this pattern weird, and define association.Initiate() which *always* initiates a carrier.

Reply:
I think that the pattern of get-or-initiate doesn't give enough guarantees of what's actually happening beneath the covers. I get that we generally don't care about the Carrier's state, but in cases in which a Carrier has constraints around how many overall messages it can handle, or rate limits, etc, it is useful to know if I'm writing the first message on a new Carrier, or tacking on my message an existing one.

How about:

- association.Initiate(), which initiates a new carrier from the association, and is essentially a way to make sure your new Carrier falls into the existing association with the right Remote and Configuration
- association.getCarriers() -> [Carrier], which accesses the list of existing Carriers if you want to see them

Reply:
That'd work for me; will try to push another commit tomorrow

Reply:
Sorry I still don't think that is what we want. You create a carrier and have a handle for it. If you don't have the handle anymore, the carrier should be closed/removed. A carrier doesn't have any long term state; that's all in association. I you need another carrier (that does the same than the thing than the previous one) then you create a new one with the same input parameters. I think you should never need to call getCarriers (expect for debugging maybe).

Reply:
okay, i buy this... agree completely that if you drop or lose a handle for a Carrier, all the underlying PSIs should go away, as should the Carrier. so will move GetCarriers() to appear along with the other debugging stuff.